src/v/resource_mgmt/available_memory.h

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/metrics.hh>
+#include <seastar/core/sstring.hh>
+
+#include <memory>
+#include <vector>
+
+namespace resources {
+
+class available_memory;
+
+struct [[nodiscard]] deregister_holder {
+    using handle_type = int;
+
+    explicit deregister_holder(available_memory* parent, handle_type handle)
+      : parent{parent}
+      , handle{handle} {}
+    deregister_holder(const deregister_holder&) = delete;
+    deregister_holder(deregister_holder&&) noexcept;
+    deregister_holder& operator=(const deregister_holder&) = delete;
+    deregister_holder& operator=(deregister_holder&&) = default;
```

Comment:
So much boilerplate just to create a move-only object.

I've wondered if it might be useful to have a trivial, empty move-only object in a utility header somewhere, then you can just add this as a member of any class: `[[no_unique_address]] move_only _mo;` which should have the same effect as all the explicit `= delete` and `= default` stuff?

Reply:
The other option is to just use `unique_ptr` though it feels ugly to do the alloc for nothing, and then you have to wrap it again if you want to hide the fact it is a unique ptr, etc.

Reply:
maybe a good use for macro. i think boost uses something like that to add these special functions in. it'd be nice if c++ had something native here.

> [[no_unique_address]] move_only _mo; which should have the same effect as all the explicit = delete and = default stuff?

probably, but clang-tidy would probably still complain

Reply:
@dotnwat - turns out boost does have [booost::noncopyable](https://www.boost.org/doc/libs/1_68_0/libs/core/doc/html/core/noncopyable.html) which seem to serve this purpose, though looking around people seem to discourage its use.

It is not to be used as a member but as a base class, which means it would benefit from EBO even without the `no_unique_address`.